homeassistant/components/xiaomi_aqara/config_flow.py

```diff
@@ -0,0 +1,167 @@
+            if self.host is not None:
+                self.selected_gateway = self.gateways.get(self.host)
+                if self.selected_gateway is not None:
+                    return await self.async_step_settings()
+
+                errors["base"] = "not_found_error"
+            else:
+                if len(self.gateways) == 1:
+                    self.selected_gateway = list(self.gateways.values())[0]
+                    return await self.async_step_settings()
+                if len(self.gateways) > 1:
+                    return await self.async_step_select()
+
+                errors["base"] = "discovery_error"
+
+        return self.async_show_form(
+            step_id="user", data_schema=GATEWAY_CONFIG, errors=errors
+        )
+
+    async def async_step_select(self, user_input=None):
+        """Handle multiple aqara gateways found."""
+        errors = {}
+        if user_input is not None:
+            ip_adress = user_input["select_ip"]
+            self.selected_gateway = self.gateways[ip_adress]
+            return await self.async_step_settings()
+
+        select_schema = vol.Schema(
+            {
+                vol.Required("select_ip"): vol.In(
+                    [gateway.ip_adress for gateway in self.gateways.values()]
+                )
+            }
+        )
+
+        return self.async_show_form(
+            step_id="select", data_schema=select_schema, errors=errors
+        )
+
+    async def async_step_zeroconf(self, discovery_info):
+        """Handle zeroconf discovery."""
+        name = discovery_info.get("name")
+        self.host = discovery_info.get("host")
+        mac_address = discovery_info.get("properties", {}).get("mac")
+
+        if not name or not self.host or not mac_address:
+            return self.async_abort(reason="not_xiaomi_aqara")
+
+        # Check if the discovered device is an xiaomi aqara gateway.
+        if not name.startswith(ZEROCONF_GATEWAY):
+            _LOGGER.debug(
+                "Xiaomi device '%s' discovered with host %s, not identified as xiaomi aqara gateway",
+                name,
+                self.host,
+            )
+            return self.async_abort(reason="not_xiaomi_aqara")
+
+        # format mac (include semicolns and make uppercase)
+        mac_address = format_mac(mac_address)
+
+        unique_id = mac_address
+        await self.async_set_unique_id(unique_id)
+        self._abort_if_unique_id_configured({CONF_HOST: self.host})
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context.update({"title_placeholders": {"name": self.host}})
+
+        return await self.async_step_user()
+
+    async def async_step_settings(self, user_input=None):
+        """Specify settings and connect aqara gateway."""
+        errors = {}
+        if user_input is not None:
+            # get all required data
+            name = user_input[CONF_NAME]
+            key = user_input.get(CONF_KEY)
```

Comment:
If the user enters an incorrect key, could that fail the connection to the gateway?

Reply:
I don't think so, keys are only used to control subdevices, but is not nessesary to get status of sensors etc.

Reply:
Still, wouldn't it be good to validate the key?

Reply:
just tested it and I can confirm that if you enter an invalid key everything will work just fine, the component will be set up and all sensors will just work and get data.
Only if you try to control something, lets say turn on the gateway light, you will get this error:
`ERROR (SyncWorker_20) [xiaomi_gateway] Got error element in data {"error":"Invalid key"}`

Reply:
@MartinHjelmare I don't think there is an easy way to test it.
If you want to test it you need to control some device, lets say toggle the gateway light. But I don't think it is nice to turn on/off the light while setting up an integration just to test...

But I could implement that if you want, simply turn on and then turn off the build in light of the gateway....

Reply:
Using `write_to_hub` seems to validate the key.

https://github.com/Danielhiversen/PyXiaomiGateway/blob/0f7ba3b53d8c72260e602feb3bcd8079d435a0c9/xiaomi_gateway/__init__.py#L330

Can we send some dummy command that doesn't really do anything but still validates the key?

Reply:
I implemented a key validation